### DISCUSSION TITLE
tracking git-head with git-refs and regex manager

### DISCUSSION THREAD

**Discussion by jasonschroeder-sfdc** [renovatebot/renovate#16980] [Request Help]
Hi, I am trying to update multiple `Dockerfile`s with content like:

```Dockerfile
FROM registry.company.com/images/teamname/foo:a1b2c3d4
```
to align with the git short sha from tip of the default branch of the internal git repo:
https://git.company.com/teamname/foo

I've got this so far, and it looks promising on the Dashboard:
```js
"regexManagers": [
    {
      "fileMatch": ["^Dockerfile$"],
      "matchStrings": [
        "^FROM registry.company.com/images/(?<orgRepo>.*?):(?<currentDigest>\\w{7}?)"
      ],
      "depNameTemplate": "https://git.company.com/{{{orgRepo}}}",
      "currentValueTemplate": "develop", // this is the default branch for some repos, but not all..
      "versioningTemplate": "git",
      "datasourceTemplate": "git-refs"
    }
  ]
```
<details>
<summary> 📎 Relevant debug-level logs from renovate</summary>
```
DEBUG: packageFiles with updates (repository=jschroeder/bar)
       "config": {
         "regex": [
           {
             "packageFile": "Dockerfile",
             "deps": [
               {
                 "depName": "https://git.company.com/teamname/foo",
                 "currentValue": "develop",
                 "currentDigest": "6f9150c",
                 "datasource": "git-refs",
                 "versioning": "git",
                 "replaceString": "FROM registry.company.com/images/teamname/foo:6f9150c",
                 "depIndex": 0,
                 "updates": [
                   {
                     "updateType": "digest",
                     "newValue": "develop",
                     "newDigest": "1f5dad384a43a759d1483b37eb1ec7d52d3c9fe3",
                     "branchName": "renovate/https-registry.company.com-teamname-foo-docker-digest"
                   }
                 ],
                 "warnings": []
               }
             ],
             "matchStrings": [
               "^FROM registry.company.com/sfci/(?<orgRepo>.*?):(?<currentDigest>\\w{7}?)"
             ],
             "depNameTemplate": "https://git.company.com/{{{orgRepo}}}",
             "currentValueTemplate": "develop",
             "datasourceTemplate": "git-refs",
             "versioningTemplate": "git"
           }
         ]
       }
```
</details>

and I end up with a mysterious warning:
```
 WARN: Error updating branch: update failure (repository=jschroeder/bar, branch=renovate/https-git.company.com-teamname-foo-docker-digest)
```

I've studied these discussions, but I'm missing something.
- https://github.com/renovatebot/renovate/discussions/8402
- https://github.com/renovatebot/renovate/discussions/16510
I have also tried empty-string `currentValueTemplate`.z

<details>
<summary>env and versions</summary>
- renovate 32.131.2
- node 14.20.0
- on-prem github enterprise
- on-prem docker registry
</details>

**Comment by rarkins** [comment]
Is it possible you can reproduce this in a public repo?

**Comment by jasonschroeder-sfdc** [comment]
Repo is now online, waiting for Renovate bot to scan it. https://github.com/jasonschroeder-sfdc/renovate-git-refs

**Comment by jasonschroeder-sfdc** [comment]
Hi @rarkins,

I've traced it down to here, in the function `confirmIfDepUpdated(...)`: https://github.com/renovatebot/renovate/blob/8a69ccfbd4ab5c75c12663357e25c1128e712c36/lib/workers/repository/update/branch/auto-replace.ts#L74-L81

It seems to be executing like this:
```js
  if ("cf2aea1" === "cf2aea1435749fe67f109b5db4b065d899a92a6d") {
    return true;
  }
  if (!"6f9150c" && !false) {
    return true;
  }
  // istanbul ignore next
  return false;
}
```

**So how do I convince the `regexManager` to use the git short sha?**

I've also tried the following in my `renovate.json` but it did not have any effect.
```json
      "extractVersionTemplate": "^(?<version>\\w{7})"
```

```json
      "extractVersionTemplate": "{{{currentDigest}}}"
```

**Reply by rarkins** [reply]
I'm not sure. We could potentially change that code to use a `startsWith()` logic to allow for different lengths but I'm not sure if that solves the problem or just moves it to another place.

**Comment by MPV** [comment]
Have you tried using `currentDigestShort`, or is it safe to assume it didn't help?
